lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp

```diff
@@ -708,549 +720,310 @@ CompilerType GetTypedefedTypeRecursive(CompilerType type) {
+    bool is_base_class = false;
+    bool is_deref_of_parent = false;
+    uint64_t language_flags = 0;
+  };
+  using GetChildInfoClosure = std::function<llvm::Expected<ChildInfo>(void)>;
+  using GetChildNameClosure = std::function<std::string(void)>;
+  /// If callback returns an error the visitor is cancelled and the
+  /// error is returned. The \c type parameter is the type of the
+  /// visited child, the \c depth parmaeter is the nesting depth.
+  using VisitCallback = std::function<llvm::Error(
+      CompilerType, unsigned, GetChildNameClosure, GetChildInfoClosure)>;
+  SwiftRuntimeTypeVisitor(SwiftLanguageRuntime &runtime, CompilerType type,
+                     ValueObject *valobj)
+    : m_runtime(runtime), m_type(type), m_valobj(valobj) {
+    if (valobj)
+      m_exe_ctx = valobj->GetExecutionContextRef();
+  }
+  SwiftRuntimeTypeVisitor(SwiftLanguageRuntime &runtime, CompilerType type,
+                          ExecutionContextScope *exe_scope,
+                          bool hide_superclass, bool include_clang_types)
+      : m_runtime(runtime), m_type(type), m_hide_superclass(hide_superclass),
+        m_include_clang_types(include_clang_types) {
+    if (exe_scope)
+      exe_scope->CalculateExecutionContext(m_exe_ctx);
+  }
+  SwiftRuntimeTypeVisitor(SwiftLanguageRuntime &runtime, CompilerType type,
+                          ExecutionContext *exe_ctx, bool hide_superclass,
+                          bool include_clang_types, bool visit_superclass)
+      : m_runtime(runtime), m_type(type), m_hide_superclass(hide_superclass),
+        m_include_clang_types(include_clang_types),
+        m_visit_superclass(visit_superclass) {
+    if (exe_ctx)
+      m_exe_ctx = *exe_ctx;
+  }
+  llvm::Error VisitAllChildren(VisitCallback callback) {
+    return VisitImpl({}, callback).takeError();
+  }
+  llvm::Expected<unsigned> CountChildren() { return VisitImpl({}, {}); }
+  llvm::Error VisitChildAtIndex(unsigned idx, VisitCallback callback) {
+    return VisitImpl(idx, callback).takeError();
+  }
+
+private:
+  /// Implements all three kinds of traversals in one function to
+  /// centralize the logic. If not counting (= visit_callback exists)
+  /// the function returns 0 on success.
+  llvm::Expected<unsigned> VisitImpl(std::optional<unsigned> visit_only,
+                                     VisitCallback visit_callback);
+};
+
+llvm::Expected<unsigned>
+SwiftRuntimeTypeVisitor::VisitImpl(std::optional<unsigned> visit_only,
+                                   VisitCallback visit_callback)
 
-  auto ts_sp = type.GetTypeSystem().dyn_cast_or_null<TypeSystemSwiftTypeRef>();
+{
+  if (!m_type)
+    return llvm::createStringError("invalid type");
+
+  const unsigned success = 0;
+  bool count_only = !visit_callback;
+  auto ts_sp =
+      m_type.GetTypeSystem().dyn_cast_or_null<TypeSystemSwiftTypeRef>();
   if (!ts_sp)
-    return llvm::createStringError("no Swift typesystem");
+    return llvm::createStringError("no type system");
```

Comment:
Also validation that could be done at construction time.

Reply:
See above.